doc/design/optimizer.md

```diff
@@ -0,0 +1,88 @@
+## Optimizer Design
+Optimizer is used to update parameter according to parameter gradient. 
+
+#### A typical training process is:
+
+1. run forward to calculate activation using data and parameter.
+1. run backward to calculate the gradient of activation and parameter using cost, activation, and parameter.
+1. apply the gradient to the corresponding parameter.
+
+User write code to describe the forward process, Optimizer should help implicitly add the process of backward and update according to forward process and other configuration.
+
+In PaddlePaddle, we use graph of operators to describe computation. So the Python class `Optimizer` defines the API to add Ops to computation block. It adds two kinds of Operators:
+
+1. Gradient Ops. Used to calculate the gradients for parameters.
+2. Optimize Ops. Used to apply gradient to parameters.
+
+The python user interface is:
+
+```python
+class Optimizer(object):
+	def _backward(loss):
+		"""
+		Add Operators to Compute gradients of `loss` 
+		It returns the backward net.
+		"""
+		...
+		return backward_net
+
+	def _update(var_list):
+		"""
+		Add Operators to Apply gradients to variables 
+		in var_list. It returns an update `Operator`.
+		"""
+		...
+		return update_op
+
+	def minimize(loss, var_list):
+		"""Add operations to minimize `loss` by updating `var_list`.
+		
+		This method simply combines calls `_backward()` and
+		`_update()`.
+		"""
+		backward_net = _backward(loss)
+		update_op = _update(backward_net)
+		return update_op
+```
+
+because we do not want users to know the step of `_backward` and `_update`, so we decide to export only `minimize()` to users.
+
+## Three situation in parameter update.
+1. One machine no card/one GPU card.
```

Comment:
There are four situations:

1. single thread or single GPU
2. multi-threads
3. multi-GPUs
4. multi-nodes

In multi-threads, we only have one copy of parameters and gradients in memory. But in multi-GPUS, we have parameters and gradients in every GPU card.

Reply:
yes, I will update this